File: procedural/queens.py
#!/usr/bin/env python

import sys;

def nqueens(size):
    row = [True] * size
    up = [True] * (2*size-1)
    down = [True] * (2*size-1)
    board = [None] * size
    def queen(n):
        if n >= size:
            print ' '.join(map(str, board))
        else:
            for i in range(0, size):
                if row[i] and up[n+i] and down[n-i]:
                   row[i] = up[n+i] = down[n-i] = False
                   board[n] = i+1
                   queen(n+1)
                   row[i] = up[n+i] = down[n-i] = True
    queen(0)

nqueens(int(sys.argv[1]))

File: procedural/queens.php
#!/usr/bin/env php
<?php
    function nqueens($size) {
        $row = array_fill_keys(range(0, $size-1), True);
        $up = array_fill_keys(range(0, 2*$size-2), True);
        $down = array_fill_keys(range(-$size+1, $size-1), True);
        $board;
        $queen = function($n) use (&$queen, &$size, &$row, &$up, &$down, &$board) {
            if ($n >= $size) {
                echo implode(' ', $board), "\n";
            } else {
                for ($i = 0; $i < $size; ++$i) {
                    if ($row[$i] && $up[$n+$i] && $down[$n-$i]) {
                        $row[$i] = $up[$n+$i] = $down[$n-$i] = False;
                        $board[$n] = $i+1;
                        $queen($n+1);
                        $row[$i] = $up[$n+$i] = $down[$n-$i] = True;
                    }
                }
            }
        };
        $queen(0);
    }
    nqueens($argv[1]);
?>
